### PR TITLE
Py3.11 sync branch with main

### PR DESCRIPTION
**Pull request by sklam**

### CODE REVIEW

**Comment by sklam** [comment]
Failing tests all seem to relate to missing llvmlite build for py3.11. I'm doing the build out now

**Comment by sklam** [comment]
For reviewers: actual changes start at https://github.com/numba/numba/pull/8814/commits/2c7926dde86450216779d077c0fdb4c581c87b98, which is the merge commit from main. The subsequent commits are for fixups.

File: buildscripts/incremental/setup_conda_environment.sh [hunk 1]
```diff
@@ -71,7 +71,7 @@ elif  [[ $(uname) == Darwin ]]; then
 fi
 
 # Install latest correct build
-$CONDA_INSTALL conda-forge::llvmdev=11
+$CONDA_INSTALL -c conda-forge numba/label/pynext_ci::llvmdev=14
```
Comment (sklam) [inline buildscripts/incremental/setup_conda_environment.sh:74]: Don't have the llvmlite buildout for py3.11 yet.

File: buildscripts/incremental/setup_conda_environment.sh [hunk 2]
```diff
@@ -27,13 +27,22 @@ source deactivate
...
+fi
+
 # If VANILLA_INSTALL is yes, then only Python, NumPy and pip are installed, this
 # is to catch tests/code paths that require an optional package and are not
 # guarding against the possibility that it does not exist in the environment.
 # Create a base env first and then add to it...
 # NOTE: gitpython is needed for CI testing to do the test slicing
 # NOTE: pyyaml is used to ensure that the Azure CI config is valid
-conda create -n $CONDA_ENV -q -y -c conda-forge ${EXTRA_CHANNELS} python=$PYTHON numpy=$NUMPY pip gitpython pyyaml
+conda create -n $CONDA_ENV -q -y -c conda-forge ${EXTRA_CHANNELS} python=$PYTHON $NUMPY_CHANNEL_PKG=$NUMPY pip gitpython pyyaml
```
Comment (stuartarchibald) [inline buildscripts/incremental/setup_conda_environment.sh:45]: Same comment as above RE use of conda-forge.

File: azure-pipelines.yml
Comment (stuartarchibald) [inline azure-pipelines.yml:1]: I'm not sure that this file correctly correlates with the changes made in `buildscripts/azure/azure-windows.yml`.
a) I think the windows/osx builds are not running (commented in this file)
b) as a result of a) the `TEST_COUNT` and `TEST_START_INDEX` enumeration is inconsistent.

File: numba/core/interpreter.py
Comment (stuartarchibald) [inline numba/core/interpreter.py:1]: `op_SETUP_EXCEPT` was removed as part of the 3.7 removal patch?

File: numba/core/byteflow.py
Comment (stuartarchibald) [inline numba/core/byteflow.py:1]: `op_SETUP_EXCEPT` was removed as part of the 3.7 removal patch?

File: numba/tests/test_parfors.py
Comment (stuartarchibald) [inline numba/tests/test_parfors.py:1]: line ~3370 "#create new code parts", comment, is this merge correct? It's removed on `main`.

File: numba/_helperlib.c
Comment (stuartarchibald) [inline numba/_helperlib.c:1]: I'm not sure the merge content in the CPP part of `reraise_exc_is_none` is correct.

File: buildscripts/incremental/test.sh [hunk 3]
```diff
@@ -112,7 +112,7 @@ python -m numba.tests.test_runtests
 echo "INFO: All discovered tests:"
 python -m numba.runtests -l
 
-# Now run the Numba test suite with slicing
+# Now run the Numba test suite with sharding
 # Note that coverage is run from the checkout dir to match the "source"
 # directive in .coveragerc
 echo "INFO: Running slice of discovered tests: ($TEST_START_INDEX,None,$TEST_COUNT)"
```
Comment (stuartarchibald) [inline buildscripts/incremental/test.sh:118]: The merge of this section doesn't seem quite right. References to "slice" etc.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch. I've reviewed up to the point of the merge commit from `main` into `py3.11`, have not checked the "fixes" as yet.

File: numba/_helperlib.c
Comment (sklam) [reply numba/_helperlib.c:1]: i thk the code is correct but the double macros make it hard to follow. i can simplify it.

File: azure-pipelines.yml
Comment (sklam) [reply azure-pipelines.yml:1]: This branch has no azure test on windows. I'm just keeping all the temporary changes on this branch.

File: numba/core/interpreter.py
Comment (sklam) [reply numba/core/interpreter.py:1]: removed in https://github.com/numba/numba/pull/8814/commits/e823aab39cc0f027720212dddb22694fcd02995a

File: numba/core/byteflow.py
Comment (sklam) [reply numba/core/byteflow.py:1]: removed in https://github.com/numba/numba/pull/8814/commits/e823aab39cc0f027720212dddb22694fcd02995a

File: numba/tests/test_parfors.py
Comment (sklam) [reply numba/tests/test_parfors.py:1]: removed in https://github.com/numba/numba/pull/8814/commits/b140a6fa1251042703da56da19b910742bade7a8

**Comment by sklam** [comment]
I have addressed everything so far. The CI buildscripts changes are intentionally left is to match what's on the py3.11 branch so the test results can be compared

File: buildscripts/incremental/setup_conda_environment.cmd [hunk 4]
```diff
@@ -14,14 +14,18 @@ cmd /C conda info
...
+
 @echo on
 
 @rem Deactivate any environment
 call deactivate
 @rem Display root environment (for debugging)
 conda list
 @rem Scipy, CFFI, jinja2 and IPython are optional dependencies, but exercised in the test suite
-conda create -n %CONDA_ENV% -q -y -c conda-forge python=%PYTHON% numpy=%NUMPY% cffi pip scipy jinja2 ipython gitpython pyyaml
+conda create -n %CONDA_ENV% -q -y -c conda-forge python=%PYTHON% %NUMPY_CHANNEL_PKG%=%NUMPY% cffi pip scipy jinja2 ipython gitpython pyyaml
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply buildscripts/incremental/setup_conda_environment.cmd:28]: ok, that sounds like a reasonable strategy.

**Comment by sklam** [comment]
See https://github.com/numba/numba/pull/8822 that merge `py3.11` to `main`

File: numba/tests/test_parfors.py
Comment (stuartarchibald) [reply numba/tests/test_parfors.py:1]: Confirmed, thanks.

File: numba/core/byteflow.py
Comment (stuartarchibald) [reply numba/core/byteflow.py:1]: Confirmed, thanks.

File: numba/core/interpreter.py
Comment (stuartarchibald) [reply numba/core/interpreter.py:1]: Confirmed, thanks.

File: azure-pipelines.yml
Comment (stuartarchibald) [reply azure-pipelines.yml:1]: I think this is a reasonable strategy, thanks for stating it. Will largely ignore CI changes as they are "temporary" with view of testing this branch etc.

File: numba/_helperlib.c
Comment (stuartarchibald) [reply numba/_helperlib.c:1]: Thanks, on inspection this now looks valid (and a lot more simple!).

**Comment by stuartarchibald** [comment]
Note: have checked 72664d4, e823aab and b140a6f.

**Comment by stuartarchibald** [comment]
Note: Patches: 2c7926dde86450216779d077c0fdb4c581c87b98..d5cc2a26a4674e9aee63ab09b934f8b415a6ca03 look ok.

**Comment by sklam** [comment]
superseded by https://github.com/numba/numba/pull/8831
